// @tamagui/web/dist/esm/helpers/themeable.native.js
import { jsx as _jsx } from "react/jsx-runtime";
import { forwardRef } from "react";
import { Theme } from "../views/Theme";
function _define_property(obj, key, value) {
  return key in obj ? Object.defineProperty(obj, key, {
    value,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : obj[key] = value, obj;
}
function _object_spread(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {}, ownKeys2 = Object.keys(source);
    typeof Object.getOwnPropertySymbols == "function" && (ownKeys2 = ownKeys2.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
      return Object.getOwnPropertyDescriptor(source, sym).enumerable;
    }))), ownKeys2.forEach(function(key) {
      _define_property(target, key, source[key]);
    });
  }
  return target;
}
function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);
  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    enumerableOnly && (symbols = symbols.filter(function(sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    })), keys.push.apply(keys, symbols);
  }
  return keys;
}
function _object_spread_props(target, source) {
  return source = source ?? {}, Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function(key) {
    Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
  }), target;
}
function _object_without_properties(source, excluded) {
  if (source == null)
    return {};
  var target = _object_without_properties_loose(source, excluded), key, i;
  if (Object.getOwnPropertySymbols) {
    var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
    for (i = 0; i < sourceSymbolKeys.length; i++)
      key = sourceSymbolKeys[i], !(excluded.indexOf(key) >= 0) && Object.prototype.propertyIsEnumerable.call(source, key) && (target[key] = source[key]);
  }
  return target;
}
function _object_without_properties_loose(source, excluded) {
  if (source == null)
    return {};
  var target = {}, sourceKeys = Object.keys(source), key, i;
  for (i = 0; i < sourceKeys.length; i++)
    key = sourceKeys[i], !(excluded.indexOf(key) >= 0) && (target[key] = source[key]);
  return target;
}
function themeable(Component, staticConfig) {
  var withThemeComponent = /* @__PURE__ */ forwardRef(function(props, ref) {
    var themeInverse = props.themeInverse, theme = props.theme, componentName = props.componentName, themeReset = props.themeReset, rest = _object_without_properties(props, [
      "themeInverse",
      "theme",
      "componentName",
      "themeReset"
    ]), element = (
      // @ts-expect-error its ok
      /* @__PURE__ */ _jsx(Component, _object_spread_props(_object_spread({
        ref
      }, rest), {
        "data-disable-theme": !0
      }))
    ), contents = /* @__PURE__ */ _jsx(Theme, {
      componentName: componentName || (staticConfig == null ? void 0 : staticConfig.componentName),
      name: theme,
      "disable-child-theme": !0,
      debug: props.debug,
      inverse: themeInverse,
      reset: themeReset,
      children: element
    });
    return contents;
  }), withTheme = withThemeComponent;
  return withTheme.displayName = "Themed(".concat((Component == null ? void 0 : Component.displayName) || (Component == null ? void 0 : Component.name) || "Anonymous", ")"), withTheme;
}
export {
  themeable
};
